Added proposal to move HTTP bridge using OpenAPI 3.1.0 (#157)
strimzi/proposals: 103-bridge-vertx5-openapi-3.1.0.md

# Move HTTP bridge to OpenAPI 3.1.0 with Vert.x 5

This proposal recommends updating the HTTP bridge in the Strimzi project to adopt OpenAPI 3.1.0, the latest minor release in the 3.x specification series.
The change is prompted by updates introduced in Vert.x as part of the upcoming upgrade from version 4.x to 5.x.

## Current situation

The Strimzi HTTP bridge uses Vert.x 4.x, consistent with all Strimzi operators.
By leveraging the `vertx-web-openapi` component, together with the underlying JSON schema validation dependency, the bridge can validate incoming HTTP requests against the OpenAPI 3.0.0 specification before processing them.

In the existing OpenAPI specification, the HTTP bridge utilizes the `nullable` keyword in some places:

* Allowing a record to be sent with a `null` value, which in Apache Kafka typically represents a "tombstone" message used in compacted topics.
* Permitting the omission of partition count and replication factor parameters when creating a topic through the Admin Client API.

> Following the `RecordValue` [definition](https://github.com/strimzi/strimzi-kafka-bridge/blob/main/src/main/resources/openapi.json#L2215) which allows the `value` field o be either a JSON object or null:

Reviewer: ```suggestion
Following the `RecordValue` [definition](https://github.com/strimzi/strimzi-kafka-bridge/blob/main/src/main/resources/openapi.json#L2215) which allows the `value` field to be either a JSON object or null:
```

## Motivation

The Vert.x team is preparing to release Vert.x 5.x, which introduces several significant changes.
One major change, particularly relevant for OpenAPI v3 support in the bridge, is the transition from the `vertx-web-openapi` component to the new `vertx-web-openapi-router` and `vertx-openapi` components.
This shift also includes an update to the underlying JSON schema validation library.

This new validation library has been refactored to align more closely with the [JSON Schema](https://json-schema.org/) specification.
Such specification defines a way for a JSON object to be null, by using the `type: null` [declaration](https://json-schema.org/understanding-json-schema/reference/null).

In contrast, the `nullable` keyword used in OpenAPI 3.0.0 is not part of the JSON Schema specification.
It is specific to OpenAPI and has no effect when using JSON Schema-compliant validators.

Due to Vert.x 5.x adhering to the JSON Schema standard, the `nullable` flag, used in the current HTTP bridge OpenAPI specification, is no longer supported.

> This change causes schema validation to fail when an HTTP client sends a record with a `null` value, even if such input was previously valid.

Reviewer: ```suggestion
This change causes schema validation to fail when an HTTP client sends a record with a `null` value, even if such input is considered valid in the previous version.
```

For more details, please refer to this [discussion](https://github.com/eclipse-vertx/vertx-json-schema/issues/144) with the maintainer of the Vert.x JSON Schema component.

## Proposal

In light of the motivation outlined above, this proposal recommends updating the HTTP bridge to support OpenAPI 3.1.0 in conjunction with the migration to Vert.x 5.x.
This upgrade enables the OpenAPI specification to fully align with the JSON Schema standard by replacing the now unsupported `nullable` flag with the `type: null` declaration.

For example, the updated `RecordValue` schema would look like this:

```json
"RecordValue": {
  "title": "RecordValue",
  "description": "Value representation for a record. It can be an array, a JSON object or a string",
  "oneOf": [
    {
      "type": "array",
      "items": {}
    },
    {
      "type": ["object", "null"]
    },
    {
      "type": "string"
    },
    {
      "type": "null"
    }
  ]
}
```

A similar change would also be required for the `NewTopic` schema definition to ensure compatibility with the updated JSON Schema validation in Vert.x 5.x.

> Of course, the OpenAPI version would also be updated within the specification itself, as shown below:

Reviewer: ```suggestion
The OpenAPI version would also be updated within the specification itself, as shown below:
```
